# Data and AI Assignment 3
![UCLL logo](https://cdn.discordapp.com/attachments/1025724154852417549/1033683765530218506/unknown.png)

Group
- Astrid Piot
- Catarina Rilho
- Catarina Afonso
- Cristina Marques
- Greg François Reynders
- Robbe Van de Velde
- Rocío Pérez Holgado
- Seppe Haems

## Foreword

This notebook is intended to be presented to the stakeholders and people who don't necessarily have a technical background.<br/>
Therefore, we will try to explain the concepts and the results in a simple way. For a more technical explanation of our modelling process, please refer to the [technical notebook](modelling.ipynb).

## Recap

Here is a recap of what we did in the previous assignment with the results we obtained and answers to the questions we asked.

### Bussiness understanding
---


#### 1) The business objectives 

A school wants to graduate as many high quality students as possible and have them occupy important positions in the workfield.<br/>
This means keeping the standard of graduates high, while also passing a lot of students.<br/>
Having talented students at their disposal makes this job easier.<br/>
So a school needs to keep its reputation high, in order to convince talented students to apply, international students and local ones.<br/>
A delicate balance between passing a lot of students, and keeping the bar high emerges.
   
#### 2) Assess situation: What needs to be done to get ready?
      
- Risk: we are not allowed to make decision's based on race, gender and disability.<br/>
- To do: gather data about students and courses. <br/>
- Cost: will be zero in terms of currency, we only need to request data.<br/>
- Constraints: time is the biggest constraint we only have a limited time before our deadlines.<br/>
- Benefits: we will predict which students have a higher succes rate, so we can actively try to get them to come to our school.
 
#### 3) Data science goals
      
- Main goal:

Given data about a student, we want to predict if they are going to be successful or not.<br/>
- Other goals:

Analyze which courses are failed/passed the most, in order to improve these courses.<br/>
How do students interact with the virtual learning environments and how does it improve their chances to succeed in the courses.<br/><br/>

### Data understanding
---

A few questions we researched in the previous assignment (see notebook from [previous assignment](../Assignment%202/assignment2.ipynb)):

##### 🔎 How many students did not submit their assessments?

There were only 173 assignments that were not submitted out of 869560 entries (0.000199%). <br/>
This is a very small percentage and we can conclude that most students do submit their assignments.

##### 🔎 Is there any correlation between disability and the student's final result?

There are 2717 disabled students (9.44 %) and 26068 non-disabled students (90.56 %) in our dataset.

<center>
<img src="images/disability_final_results_distribution.png" style="max-width: 90%;"/><br/>
</center>

Disabled students have higher percentages of "Fail" and "Withdrawn" than not disabled students.

##### 🔎 Is there any correlation between the gender and the student's final result?

There are 13396 male students (51.82%), and 12455 female students (48.18%) in our dataset.<br/>
There is a 0.4 % difference between the female and male students' average scores, but we are not sure if this is statistically significant.

<center>
<img src="images/gender_scores.png" /><br/>
</center>

We also observed a weird pattern: In all of the scores, except at 1, 20, 40, 60, 80 and 100, there seem to be more males than females, but at those values there are considerably more females. We don't know if this is a mistake, or if female students are being evaluated differently than male students.

<center>
<img src="images/gender_final_results.png" style="max-width: 90%;"/><br/>
</center>

From this data, we concluded that there are no big differences between the two genders' final results.

##### 🔎 Is there any correlation between the IMD score and the student's final result?
<blockquote> 
The IMD score is a measure of the deprivation of an area, and it ranges from 1 to 10, where 1 is the most deprived area and 10 is the least deprived area.
</blockquote>
<br/>
<center>
<img src="images/imd_scores.png" style="max-width: 90%; max-height: 550px;"/><br/>
</center>

We observed a positive correlation between the IMD score and scores of the students, which means the higher the IMD, the higher the average score of the students.

<center>
<img src="images/imd_final_results.png" style="max-width: 90%;"/><br/>
</center>

There is also a higher percentage of students who pass or have a distinction on the higher imd_band levels, and a higher percentage of failed students on the lower levels.


##### 🔎 Is there any correlation between age and the student's final result?

There are 20665 students who are 35 years old or younger, 8298 students between the ages of 35 and 55, and only 184 older than 55 years old. This is a very big difference, so drawing conclusions from this data might be misleading.

<center>
<img src="images/age_scores.png" style="max-height: 600px;"/><br/>
</center>

Still, we can see that the average score of the students increases with age, which means that older students have higher scores on average.

##### 🔎 Is there any correlation between the highest education level and the student's final result?

<center>
<img src="images/education_final_results.png" style="max-height: 600px;"/><br/>
</center>

We can see that the percentage of students who pass or have a distinction increases with the highest education level, and the percentage of failed students decreases.

##### 🔎 Is there any correlation between the region and the student's final result?

First we noticed that a lot of column values were missing for students from the Scotland region. Our approach was to notify the business about it.

<center>
<img src="images/region_final_results.png" style="max-height: 600px;"/><br/>
</center>

The East Midlands region has the highest percentage of students who Widthdraw.<br/>
The North Western region has the highest percentage of Failed students.<br/>

Ireland has the highest percentage of passed students, and the North Region has the highest percentage of students with a distinction.<br/>

If we look at the sum of "Passed" and "Distinction" as being "Successful" and the sum of being "Failed" or "Widthdrawn" as "Unsuccessful", then we can see that the North region is in the lead with 75% successful students and the North Western Region is at the bottom with 67% of successful students.<br/>

We concluded that the region does have an impact on the student's final result.

##### 🔎 Is there any correlation between interaction with the VLE's and the student's final result?

It was difficult to answer this question because there were a lot of outliers. <br/>

<center>
<img src="images/sum_clicks_final_results.png" style="max-height: 600px;"/><br/>
</center>

We don't know if this is statistically significant, but we can conclude that there is a correlation between the interaction with the VLE and the student's final result. Students who pass or have a distinction have a higher average number of clicks than students who fail or withdraw.

### Data quality
---

- Dates and other columns were in weird formats, and we didn't know how to interpret them.
- There were a lot of missing values, especially on rows of students from the Scotland region.

We contacted the business about these issues, and a more complete dictionary was provided as well as a dataset with the missing values for the Scotland region filled in.<br/>
Unfortunately, we didn't have enough time to do a full analysis of the new dataset. We did however use it for the modelling part of this assignment.

## Data Modelling

### 🎯 Prediction target: Final result

The main goal of this project was to predict the **final result** for a new student, using their data.<br/>

We merged the categories "Distinction" and "Pass" into one category since they are both considered successful results.<br/>
The category "Withdrawn" was not considered because it is not a final result, and we don't know if the student would eventually pass or fail.<br/>

So our prediction target is if a student will "Pass" or "Fail".

### 📊 Features

For a new student, we wouldn't have data on their assessment scores or interactions with the VLE, so we decided not to include these features.<br/>
The most relevant data we would have on a new student would be in the "Student Info" table.<br/>

Given the time constraints, we decided to use only the features that we analyzed in the previous assignments and that we thought were relevant to the prediction.<br/>
These features are: **gender, region, highest education, IMD score, age, number of previous attempts, studied credits and disability**.

If we had more time, we would have also investigated if the module or the presentation were relevant to the prediction, since taking different modules or registering at a different time of the year might impact the student's final result.<br/>
But for now we did not include these features.

### 📈 Model selection

Our target variable is the final results which can be "Pass" or "Fail". This is a binary classification problem, so we decided to use classification models.<br/>
We tried the following models: **Naive Bayes, Logistic Regression, Decision Tree, Random Forest, KNN, SVM, Neural Network**.<br/>

We used different metrics to evaluate the models, such as:
- **Accuracy** - how many times the model was correct overall.
- **Precision** - the ratio of correctly predicted positive observations to the total predicted positive observations.
- **Recall** - how many of the positive observations in a dataset the model is able to correctly identify.


We also used cross validation to verify that the models were not overfitting.<br/>

After comparing the different metrics of the models, we decided that the best model for our use-case was the Decision Tree model.<br/>
It had an accuracy of 68,13%, a precision of 50,36% and a recall of 20,28%.<br/>

Even though our Neural Network model had a sligthly better accuracy (68,57%), it took a lot longer to run and the recall was worse, so we decided to go with the Decision Tree model.

### ✅ Model evaluation

![Decision Tree Confusion matrix](images/decision_tree_confusion_matrix.png)

In [1]:
print("Students who actually passed: ", 5317 + 551)
print("Students who actually failed: ", 2198 + 559)
print("Students who were predicted to pass: ", 5317 + 2198)
print("Students who were predicted to fail: ", 551 + 559)

Students who actually passed:  5868
Students who actually failed:  2757
Students who were predicted to pass:  7515
Students who were predicted to fail:  1110


In the data there are 5868 students that pass and 2757 that failed.
In the confusion matrix we can see that the model predicted that 7515 students would pass, but of these 7515 only 5317 actually passed. On other hand, the model predicted that 1110 students would fail, but only 559 did. 

Looking at these results, we can see that even though the accuracy of this model stands over 68.13%, it makes a lot more mistakes when is dealing with failed students.

This is probably because in our dataset there are a lot more students that pass, so even if the model guesses that every student passes, the accuracy would still be high.

Every model that we made always predicted that most students would pass. So that is why we looked for a model that would have a better recall for failed students.

We can also have a quick look at how the decision tree makes it predictions. We won't go into detail but as an example we can look at the top 3 levels of the tree:

![Decision Tree](images/decision_tree_3_levels.png)

We can see here that the model thinks that the most important feature is the number of previous attempts. If the student has less than 0.5 attempts (so 0) it will look at the student's highest education. 
After that it will look at the age of the student. We can see that this model thinks that number of previous attempts, age and imd score are the most important features to predict if a student will pass or fail.

So using this model to predict if a new student will pass or fail we would get around 68% of the cases correct , but it will most likely predict that the student will pass. 

An advantage of using a decision tree is also that it is very fast so it takes less resources than many other models. It's also a very intuitive model because it's easy to read and visualize everything.